File: single-link/single-link.cpp
/*####################################################################
###################  Algoritmo single-link	########################
####################################################################*/
						
#include <iostream>
#include <math.h>
#include <vector>

using namespace std;

typedef struct{
	string nome;
	double d1;
	double d2;
} objeto;

typedef struct{
	vector<objeto> objetos;
}cluster;


//calcula a distância euclidiana entre dois clusters com n objetos cada. Cada objeto com dois atributos
double distanciaEuclidiana(cluster c1, cluster c2){
	double menorDist = (double)INT32_MAX;
	for(int i = 0; i < c1.objetos.size(); i++){
		for(int j = 0; j < c2.objetos.size(); j++){
			double d = pow((c1.objetos[i].d1-c2.objetos[j].d1), 2) + pow((c2.objetos[j].d2-c1.objetos[i].d2), 2);
			if(d < menorDist)
				menorDist = d;
		}
	} 
	return sqrt(menorDist); 
}

int main(){
	vector<cluster> clusters;

	cluster c;
	objeto p;
	// recebe objetos e seus atributos
	while(cin >> p.nome){
		cin >> p.d1 >> p.d2;

		c.objetos.push_back(p);
		clusters.push_back(c);
		c.objetos.clear();
	}

	int kmin = 5, kmax = 12;
	while(clusters.size() > kmin){
		double menorDist = (double)INT32_MAX;
		int menor_i, menor_j;
		double m;

		// calcula distancias e guarda a menor entre dois objetos/clusters
		for(int i = 0; i < clusters.size(); i++){
			for(int j = 0; j < i; j++){
				m = distanciaEuclidiana(clusters[i], clusters[j]);

				if(m < menorDist){
					menorDist = m;
					menor_i = i;
					menor_j = j;
				}

				if(m == menorDist && clusters[i].objetos.size()+clusters[j].objetos.size() > clusters[menor_i].objetos.size()+clusters[menor_j].objetos.size()){
					menorDist = m;
					menor_i = i;
					menor_j = j;
				}
			}
		}

		// atualiza clusters juntando os dois com menor distancia entre eles
		for(int i = 0; i < clusters[menor_j].objetos.size(); i++){
			clusters[menor_i].objetos.push_back(clusters[menor_j].objetos[i]);
		}
		clusters.erase(clusters.begin()+menor_j);		

		// imprime o conteudo de todos os clusters
		if(clusters.size() <= kmax){
			for(int i = 0; i < clusters.size(); i++){
				for(int j = 0; j < clusters[i].objetos.size(); j++)
					cout << clusters[i].objetos[j].nome << " ";
				cout << endl; 
			}
			cout << endl;
		}
	}


	return 0;
}

File: single-link/single-link.py
####################################################################
###################    Algoritmo single-link	####################
####################################################################
import sys
import math
import random

sys.path.insert(1, '../')

from util import plot, salvar

class Objeto:

	# Parâmetros: nome, lista de coordenadas
	def __init__(self, nome, coordenadas):
		self.nome = nome
		self.coordenadas = coordenadas

	# calcula a distância euclidiana do objeto até outro objeto (obs: eles devem possuir o mesmo número de coordenadas)
	# Fazer uma exceção pra caso não tenham???
	def distanciaEuclidiana(self, objeto):
		soma = 0
		
		for i in range (0, len(self.coordenadas)):
			soma += math.pow((objeto.coordenadas[i] - self.coordenadas[i]), 2)

		return math.sqrt(soma)

class Cluster:

	def __init__(self, numCoordenadas, objetos=[]):
		self.objetos = objetos
		self.numCoordenadas = numCoordenadas

	def adicionaObjeto(self, objeto):
		self.objetos.append(objeto)

	def adicionaCluster(self, cluster):
		for cl in cluster.objetos:
			self.objetos.append(cl)

	def apaga(self):
		self.objetos.clear()

def single_link(listaClusters, k_min, k_max, nomeArquivoSaida):

	numGraficos = 0
	numClusters = len(listaClusters)

	# Enquanto o número de clusters for maior que o k_min
	while numClusters > k_min:

		menor = (2**32) - 1
		soma = 0

		# Calcula todas as distâncias e acha a menor
		for i in range(0, numClusters):
			for objA in listaClusters[i].objetos:
				
				for j in range(i+1, numClusters):
					for objB in listaClusters[j].objetos:
						
						dist = objA.distanciaEuclidiana(objB)
						soma += 1
						if dist < menor:
							menor = dist
							min_indice = i
							max_indice = j

		print(soma)

		# Agora, devemos "linkar" os clusters, passando todos os clusters de um para o outro e eliminando o que ficou sem clusters
		listaClusters[min_indice].adicionaCluster(listaClusters[max_indice])
		listaClusters.pop(max_indice)
		numClusters = numClusters - 1

		if numClusters <= k_max:
			numGraficos += 1
			plot(numGraficos, "graficos", nomeArquivoSaida + str(numClusters), listaClusters)
			salvar("saidas", nomeArquivoSaida + str(numClusters), listaClusters)

		# print(numClusters)

	return listaClusters # Tem que mudar isso pra uma lista

####################################################################
############################	MAIN	############################
####################################################################

# input("Nome do arquivo de entrada: ")
nomeArquivo = sys.argv[1]
nomeArquivoSaida = sys.argv[2]
k_min = int(sys.argv[3])
k_max = int(sys.argv[4])
listaClusters = []

F = open(nomeArquivo, "r")
F.readline()
string = F.readline()

# Lê um arquivo de entrada contendo o nome do objeto e suas coordenadas
while string != "":
	valores = string.split("\t");
	listaClusters.append(Cluster(2, [Objeto(str((valores[0])), [float(valores[1]), float(valores[2])])]))
	string = F.readline()

F.close()

listaClusters = single_link(listaClusters, k_min, k_max, nomeArquivoSaida)